File: its-my-training-day/leetcode-top-75/questions/strings/stack/valid-parentheses/solution.ts
import {
  ArrayList,
  Stack,
  Queue,
  ListNode,
  TreeNode,
  MinHeap,
  MaxHeap,
} from "../../../../shared/typescript/ds";

export const validParentheses = (s: string): boolean => {
  const stack = new Stack<string>();
  const pairs: Record<string, string> = { ")": "(", "]": "[", "}": "{" };
  for (let i = 0; i < s.length; i += 1) {
    const ch = s[i];
    if (pairs[ch]) {
      const top = stack.pop();
      if (top === null || top !== pairs[ch]) return false;
    } else {
      stack.push(ch);
    }
  }
  return stack.isEmpty();
};

File: its-my-training-day/leetcode-top-75/questions/strings/stack/valid-parentheses/solution.cpp
#include "../../../../shared/cpp/ds.hpp"

#include <algorithm>
#include <cctype>
#include <cstdint>
#include <functional>
#include <string>
#include <unordered_map>
#include <vector>

bool validParentheses(const std::string& s) {
    Stack<char> stack;
    std::unordered_map<char, char> pairs;
    pairs[')'] = '(';
    pairs[']'] = '[';
    pairs['}'] = '{';
    for (char ch : s) {
        if (pairs.find(ch) != pairs.end()) {
            char top = stack.pop();
            if (top == 0 || top != pairs[ch]) return false;
        } else {
            stack.push(ch);
        }
    }
    return stack.empty();
}
